Support camelCase CSV output argument for file_name and save_interval.

JosephWoodward opened the pull request:
This change adds support for camelCased versions of file_name and save_interval CSV output arguments,
writing a warning to the logs informing users the older snake cased argument will be deprecated in time.

<!--


  (ﾉ◕ヮ◕)ﾉ*:・ﾟ✧
  
  Thank you for your interest in contributing to the k6 project!
  
  Before you get started, we'd kindly like to ask you to read our:
    - Contribution guidelines at https://github.com/grafana/k6/blob/master/CONTRIBUTING.md
    - Code of Conduct at https://github.com/grafana/k6/blob/master/CODE_OF_CONDUCT.md
    
  Out of respect for your time, please start a discussion regarding any bigger contributions either
  in a GitHub Issue, in the community forums or in the #contributors channel of the k6 slack before you
  get started on the implementation.
  
  If you've already done all of that, you're more than welcome to proceed with your pull request.
  Thank you again for your contribution! 🙏🏼
  
  
-->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/k6?pullRequest=2150) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/k6?pullRequest=2150) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/k6?pullRequest=2150) it.</sub>

JosephWoodward commented on output/csv/config.go:
```diff
@@ -59,7 +60,7 @@ func (c Config) Apply(cfg Config) Config {
 
 // ParseArg takes an arg string and converts it to a config
-func ParseArg(arg string) (Config, error) {
+func ParseArg(arg string, logger *logrus.Logger) (Config, error) {
```
Do you have any thoughts on how we could avoid such a breaking change to the package's pubic API? Other options I was considering was:

1. Use `Config` to indicate a soon to be deprecated argument was used, and write the log message further up the call stack where the logger is initialised (in `output.go`)
2. Move the checking of the soon to be deprecated arguments to `output.go`

imiric replied:
Hi, thanks for the PR!

Hhmm yeah, this is not ideal, but I think I prefer leaving it as is more than the alternatives, as currently it's clear that the warning happens where the configuration is assembled, and doing that validation outside of `config.go` would muddy things a bit.

Other than that a package like [`warnings`](https://pkg.go.dev/gopkg.in/warnings.v0) might help, though it's not very idiomatic and we wouldn't want to add another dependency because of this.

So I'd say to leave it as is, unless someone from @grafana/k6-devs has a better idea.

JosephWoodward replied:
Okay, thanks @imiric 

> Hhmm yeah, this is not ideal, but I think I prefer leaving it as is more than the alternatives, as currently it's clear that the warning happens where the configuration is assembled, and doing that validation outside of config.go would muddy things a bit.

Yeah, I agree - this way keeps the logic local. Interested to see what @grafana/k6-devs think.

yorugac replied:
Perhaps, it makes sense to add this case (logging during config parsing) to #883 as "nice-to-have" to figure out how best to approach this if it comes up again?

JosephWoodward replied:
@yorugac Good idea, I've just added this as a nice to have to that issue. :+1:

JosephWoodward commented on output/csv/config.go:
```diff
@@ -92,7 +99,8 @@ func ParseArg(arg string) (Config, error) {
 // GetConsolidatedConfig combines {default config values + JSON config +
 // environment vars + arg config values}, and returns the final result.
-func GetConsolidatedConfig(jsonRawConf json.RawMessage, env map[string]string, arg string) (Config, error) {
+//nolint: lll
```
Passing the additional logger parameter causes the line length linting error, but given we're only passing the logger until people have migrated to the camel case version I felt the no lint comment would be sufficient for the time being.

MStoykov replied:
I in general prefer to split the `func <name>(<arguments>) (<results>){` on three lines as in 
```
func <name>(
    <arguments>,
)(<results>){ 
```
as it keeps it more readable especially when lines start to wrap.

MStoykov commented on output/csv/config_test.go:
```diff
@@ -91,23 +94,37 @@ func TestParseArg(t *testing.T) {
 
 		t.Run(arg, func(t *testing.T) {
-			config, err := ParseArg(arg)
+			config, err := ParseArg(arg, testLog)
```
This could've also tested that the warning is being logged

MStoykov commented on output/csv/config.go:
```diff
@@ -76,11 +77,17 @@ func ParseArg(arg string) (Config, error) {
 		}
 		switch r[0] {
 		case "save_interval":
+			logger.Warnf("CSV output argument '%s' will soon be deprecated, please use 'saveInterval' instead.", r[0])
```
```suggestion
			logger.Warnf("CSV output argument '%s' is deprecated, please use 'saveInterval' instead.", r[0])
```
It is deprecated, we just still support it. As I don't know if we will ever remove it (maybe when we redo configuration) I don't think it's needed to also add "we will remove it in v0.XX.0" or something like that that but it's possible I guess

JosephWoodward commented:
@MStoykov Thank you for the comments, I will get these sorted this evening.

JosephWoodward commented:
@MStoykov This is ready for a re-review when you can please. I've added an assertion over the expected log messages and level, and updated your other comments too.

inancgumus reviewed (approved):
LGTM

JosephWoodward commented:
@FloorD That'd be great! Big fan of what everyone is doing at K6/Grafana so certainly wouldn't say no!
